### ISSUE TITLE
[Feature Request] `fav --version` shows commit hash

### ISSUE THREAD

**Issue by yuioto** [kingwingfly/fav#9]
Showing compilation information when `fav -V` will help report bugs.

In addition, you can also consider creating an issue template.

**Comment by kingwingfly** [comment]
I just added issue templates.

**Comment by kingwingfly** [comment]
I'll add commit hash for `-V` once my finishing refactoring.

**Comment by yuioto** [comment]
This commit (v0.2.16 53e9f8f35380b527b36056ae2d9316706b653a96) don't work.

## reproduce

```shell
cargo build --release --forzen --all-features
cd target/release
./fav -V
```
I get output.

```log
fav_cli v0.2.16
RUSTC: x86_64-unknown-linux-gnu stable 1.78.0
```

## expected

output like rustc.

```log
fav_cli v0.2.16 (53e9f8f35 2025-05-10) (Arch Linux ...)
...
commit-hash: 53e9f8f35380b527b36056ae2d9316706b653a96
...
```

The unimportant parts are omitted by `...`

## rustc

```log
rustc 1.78.0 (9b00956e5 2024-04-29) (Arch Linux rust 1:1.78.0-1)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: x86_64-unknown-linux-gnu
release: 1.78.0
LLVM version: 17.0.6
```

**Comment by yuioto** [comment]
45c750d4f73343db697fe8bd13c1a353dea77295 也是有问题，release里写了支持，但是没发现哪个commit是work的。

```log
reimu@reimu /tmp/fav (git)-[tags/v0.2.16] % git checkout 45c750d4f73343db697fe8bd13c1a353dea77295
之前的 HEAD 位置是 53e9f8f fix: action
HEAD 目前位于 45c750d hash version
reimu@reimu /tmp/fav (git)-[v0.2.16~5] % cargo build --release --frozen --all-features        
   Compiling fav_cli v0.2.16 (/tmp/fav/fav_cli)
   Compiling fav v0.2.16 (/tmp/fav/fav)
    Finished `release` profile [optimized] target(s) in 2m 58s
cargo build --release --frozen --all-features  174.08s user 2.06s system 98% cpu 2:58.20 total
reimu@reimu /tmp/fav (git)-[v0.2.16~5] % cd target/release                                    
reimu@reimu /tmp/fav/target/release (git)-[v0.2.16~5] % ./fav -V                                             
fav_cli v0.2.15-8-g45c750d
RUSTC: x86_64-unknown-linux-gnu stable 1.78.0
```

**Comment by kingwingfly** [comment]
Thank you for your testing!!!

This is because v0.2.16(53e9f8f) has a tag, so it won't show commit hash. 

If I make a new commit:
```
git log
commit 79cc59336d05ab0c87e3d6b78447521ade1ef353 (HEAD -> dev)
Author: Louis <836250617@qq.com>
Date:   Sun May 12 07:14:12 2024 +0800

    version: add timestamp

commit 53e9f8f35380b527b36056ae2d9316706b653a96 (tag: v0.2.16, origin/dev)
Author: Louis <836250617@qq.com>
Date:   Fri May 10 13:59:03 2024 +0800

    fix: action
```
Then:
```
cargo run -- -V
...
fav_cli v0.2.16-1-g79cc593 2024-05-12T07:14:12.000000000+08:00
RUSTC: aarch64-apple-darwin stable 1.77.2
```
`v0.2.16-1-g79cc593` mean "1 commit after v0.2.16 and commit hash is 79cc593".

I also tested 45c750d
```
git branch
* (HEAD detached at 45c750d)
  dev

cargo run -- -V

fav_cli v0.2.15-8-g45c750d
RUSTC: aarch64-apple-darwin stable 1.77.2
```
Works perfect. 

In your example:
> reimu@reimu /tmp/fav (git)-[tags/v0.2.16] % git checkout 45c750d4f73343db697fe8bd13c1a353dea77295
> ```                          
> fav_cli v0.2.15-8-g45c750d
>                    ^^^^^^^ here it is
> ```

**Comment by yuioto** [comment]
I'm sorry, I misread it.

I found a new problem.

If my build directory is not a git repository.

```log
reimu@reimu ~/data/work/tmp % cargo build
...
   Compiling clap_complete v4.5.2
warning: fav_cli@0.2.16: not within a suitable 'git' worktree!
warning: fav_cli@0.2.16: VERGEN_GIT_DESCRIBE set to default
...
reimu@reimu ~/data/work/tmp/target/debug % ./fav -Vv
fav_cli VERGEN_IDEMPOTENT_OUTPUT
RUSTC: x86_64-unknown-linux-gnu stable 1.78.0
```
## expected

Use a method to indicate that it is before or after a tag.

for example:

```log
fav_cli v0.1.15+
RUSTC: x86_64-unknown-linux-gnu stable 1.78.0
```

**Comment by yuioto** [comment]
This may be a bit difficult, so as an alternative:

```log
fav_cli v0.1.15(cargo tag)
RUSTC: x86_64-unknown-linux-gnu stable 1.78.0
```

**Comment by kingwingfly** [comment]
fixed on d43adf9
```
rm -rf .git
cargo run -- -V
   Compiling fav_cli v0.2.17 (/Users/louis/rust/fav_test/fav_cli)
   Compiling fav v0.2.17 (/Users/louis/rust/fav_test/fav)
    Finished dev [unoptimized + debuginfo] target(s) in 2.33s
     Running `/Users/louis/rust/.cargo/target/debug/fav -V`
fav_cli 0.2.17+ unknown-timestamp
RUSCT aarch64-apple-darwin stable 1.77.2
```
I see a spell fault `RUSCT`, hahaha

**Comment by yuioto** [comment]
我之所以说这是困难的是因为使用`v0.0.0+`这种写法会造成歧义。
因为如果版本是`v0.0.0`他依然会写
`v0.0.0+`，这会引起误解，使用[上一条](https://github.com/kingwingfly/fav/issues/9#issuecomment-2106082864)中写的或`v0.0.0~v0.0.0+`的写法会更好。
（从mod.rs#13看到只是加了一个+，如果我错了，抱歉。）

---

The reason I say this is difficult is because using `v0.0.0+` will cause ambiguity.
Because if the version is `v0.0.0`, it will still be written
`v0.0.0+`, which will cause misunderstanding. It is better to use the writing in [the previous article](https://github.com/kingwingfly/fav/issues/9#issuecomment-2106082864) or `v0.0.0~v0.0.0+`.
(From mod.rs#13, I saw that only a + was added. Sorry if I am wrong.)
